feat(dx): additional log to help locate a potential missing provider

mrossard opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | main 
| Tickets       | Closes #6335
| License       | MIT

Just adding a simple log to avoid completely hiding the fact that there was a ProviderNotFoundException behind a NotFoundHttpException. 
It could help save time for devs who simply forgot to implement the provider.

mrossard commented:
Am I missing something? I only added 3 lines to a file, what's the conflict here? :D

dunglas commented:
@mrossard the conflict is in `use`. You need to rebase.

dunglas commented on src/Symfony/EventListener/ReadListener.php:
```diff
@@ -50,6 +51,7 @@ public function __construct(
         ?ResourceMetadataCollectionFactoryInterface $resourceMetadataCollectionFactory = null,
         private readonly ?SerializerContextBuilderInterface $serializerContextBuilder = null,
         LegacyUriVariablesConverterInterface|UriVariablesConverterInterface|null $uriVariablesConverter = null,
+        private readonly LoggerInterface $logger
```
```suggestion
        private readonly ?LoggerInterface $logger = null,
```

(To prevent a BC break).

mrossard replied:
sure, done.

dunglas commented on src/Symfony/EventListener/ReadListener.php:
```diff
@@ -142,6 +144,8 @@ public function onKernelRequest(RequestEvent $event): void
                 || ($operation instanceof Put && !($operation->getAllowCreate() ?? false))
             )
         ) {
+            //Log it in debug in case the dev just forgot to implement it
```
```suggestion
            // In case the dev just forgot to implement it
```

mrossard replied:
done.

dunglas commented on src/Symfony/EventListener/ReadListener.php:
```diff
@@ -142,6 +144,8 @@ public function onKernelRequest(RequestEvent $event): void
             )
         ) {
+            //Log it in debug in case the dev just forgot to implement it
+            $this->logger->debug("No provider found for " . $resourceClass);
```
```suggestion
            $this->logger->debug('No provider registered for {resource_class}', ['resource_class' => $resourceClass]);
```

mrossard replied:
done.

mrossard commented:
> @mrossard the conflict is in `use`. You need to rebase.

oh, cool, i'm just dumb then. :D I was sure i had pulled the changes before creating the branch, sorry.

I'll try to look at how it can be added in the profiler/ web debug toolbar too, that's a great point.

Aerendir commented on src/Symfony/EventListener/ReadListener.php:
```diff
@@ -142,6 +144,8 @@ public function onKernelRequest(RequestEvent $event): void
         ) {
+            //Log it in debug in case the dev just forgot to implement it
+            $this->logger->debug("No provider found for " . $resourceClass);
             throw new NotFoundHttpException('Not Found');
```
Maybe it could be useful to also pass the previous exception here 🙏

mrossard replied:
I could add the exception message or even the stacktrace, but that seems a bit overkill doesn't it? Having a log saying something like this should be fine, shouldn't it?

`app.DEBUG: No provider registered for App\ApiResource\TestResource {"resource_class":"App\\ApiResource\\TestResource"} []`

mrossard replied:
sorry, misunderstood what you meant. I'll do that.

Aerendir replied:
@mrossard , thank you! 🙏

mrossard commented:
I rebased it and added the logger to the api_platform.state_provider.read service definition on the symfony config side. Not sure if there's something similar to do for laravel...?

soyuka commented on src/State/Provider/ReadProvider.php:
```diff
@@ -84,6 +88,8 @@ public function provide(Operation $operation, array $uriVariables = [], array $c
             )
         ) {
+            // In case the dev just forgot to implement it
+            $this->logger?->debug('No provider registered for {resource_class}', ['resource_class' => $resourceClass]);
```
imo this line should be at line :81

mrossard replied:
I wasn't sure that was something useful to report for PUT and POST, but that's easy enough to change.

soyuka commented on src/State/Provider/ReadProvider.php:
```diff
@@ -84,6 +88,8 @@ public function provide(Operation $operation, array $uriVariables = [], array $c
         ) {
+            // In case the dev just forgot to implement it
+            $this->logger?->debug('No provider registered for {resource_class}', ['resource_class' => $resourceClass]);
             throw new NotFoundHttpException('Not Found');
```
we could also keep the previous exception (ProviderNotFound) and use it here as `previous` argument of `NotFoundHttpException`

mrossard replied:
That's a clear benefit of moving it to line 81. I'll do that.

Aerendir reviewed (approved):
Looks good on my side. Thank you! 🙏
